### PR TITLE
TST: add test for .unique() dtype preserving

### LINKED ISSUE #15442: COMPAT: .unique() should be type preserving
xref https://github.com/pandas-dev/pandas/pull/15439/files#r101811209

```
In [1]: Series([1, 2, 2], dtype='int8').unique()
Out[1]: array([1, 2])

In [3]: Series([1, 2, 2], dtype='int32').unique()
Out[3]: array([1, 2])

In [4]: Series([1, 2, 2], dtype='int64').unique()
Out[4]: array([1, 2])

In [5]: Series([1, 2, 2], dtype='float32').unique()
Out[5]: array([ 1.,  2.])
```

these should all be the original dtype.

We only create hashtables (and use them in unique) for ``int64``, ``float64`` and ``uint64``, should do this for other dtypes. 
https://github.com/pandas-dev/pandas/blob/master/pandas/src/hashtable_class_helper.pxi.in#L222

I thought there was an issue about this, but guess not.

### PR DESCRIPTION
**Pull request by ganevgv**
- [x] closes #15442
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/tests/series/test_dtypes.py [hunk 1]
```diff
@@ -506,6 +506,33 @@ def test_is_homogeneous_type(self):
         assert Series([1, 2])._is_homogeneous_type
         assert Series(pd.Categorical([1, 2]))._is_homogeneous_type
 
+    @pytest.mark.parametrize(
```
Comment (jreback) [inline pandas/tests/series/test_dtypes.py]: this should go in pandas/tests/test_algos.py.

File: pandas/tests/series/test_dtypes.py [hunk 2]
```diff
@@ -506,6 +506,33 @@ def test_is_homogeneous_type(self):
...
+            ([1, 2, 2], [1, 2], "uint8"),
+            ([1, 2, 2], [1, 2], "uint16"),
+            ([1, 2, 2], [1, 2], "uint32"),
+            ([1, 2, 2], [1, 2], "uint64"),
+            ([1, 2, 2], [1.0, 2.0], "float16"),
+            ([1, 2, 2], [1.0, 2.0], "float32"),
+            ([1, 2, 2], [1.0, 2.0], "float64"),
+            ([1, 2, 2], [1.0, 2.0], "complex64"),
+            ([1, 2, 2], [1.0, 2.0], "complex128"),
+            ([True, True, False], [True, False], "bool"),
```
Comment (jreback) [inline pandas/tests/series/test_dtypes.py]: I would really try to use the existing fixtures for this, e.g. any_numpy_dtype (and just cast the input in numpy before construction), skipping those which don't make sense.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/series/test_dtypes.py (hunk 1 above)
Comment (ganevgv) [reply pandas/tests/series/test_dtypes.py]: Thanks, @jreback - moved the test there.

File: pandas/tests/series/test_dtypes.py (hunk 2 above)
Comment (ganevgv) [reply pandas/tests/series/test_dtypes.py]: Thanks, @jreback - now using `np.sctypes` to extract all dtypes and skipping `[bytes, str, np.void]`.

File: pandas/tests/test_algos.py [hunk 3]
```diff
@@ -353,6 +353,25 @@ def test_on_index_object(self):
 
         tm.assert_almost_equal(result, expected)
 
+    @pytest.mark.parametrize(
+        "data, uniques, dtype_list",
+        [
+            ([1, 2, 2], [1, 2], np.sctypes["int"]),
```
Comment (jreback) [inline pandas/tests/test_algos.py]: i would rather not use numpy references here and instead use our fixtures or definitions in pandas/conftest.py they are much more comprehensive and include pandas types

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/test_algos.py (hunk 3 above)
Comment (ganevgv) [reply pandas/tests/test_algos.py]: Thanks for the feedback, @jreback - now using `any_numpy_dtype` to extract the dtypes.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @ganevgv

**Pull request merged**